lib/ui/volatile_path_tracker.cc

```diff
@@ -0,0 +1,42 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/volatile_path_tracker.h"
+
+namespace flutter {
+
+VolatilePathTracker::VolatilePathTracker(
+    fml::RefPtr<fml::TaskRunner> ui_task_runner)
+    : ui_task_runner_(ui_task_runner) {}
+
+void VolatilePathTracker::insert(std::shared_ptr<Path> path) {
+  FML_DCHECK(ui_task_runner_->RunsTasksOnCurrentThread());
+  FML_DCHECK(path);
+  FML_DCHECK(path->path_.isVolatile());
+  paths_.insert(path);
+}
+
+void VolatilePathTracker::erase(std::shared_ptr<Path> path) {
+  FML_DCHECK(path);
+  fml::TaskRunner::RunNowOrPostTask(ui_task_runner_,
```

Comment:
I wonder if `fml::TaskRunner::RunNowOrPostTask`'s overhead could be significant here. Can we add a benchmark that inserts/erases 10,000 simple paths (e.g., tiny rects) per frame?

Reply:
This will get called from whatever thread Dart is doing GC on. In local testing, that's working out to be a random worker thread from the VM. 

I'm not sure what it is you'd like to benchmark - speed of erasure when a different thread is actually used? 

The only other option I can think of for this is to use a mutex to guard the collection, but that should be slower than this method since it would need to acquire the lock in all three methods rather than just jumping the thread in this one.

Reply:
Another possibility would be a batching mechanism that schedules a deletion task with a delay and deletes all paths that were accumulated during the delay (similar to how `SkiaUnrefQueue` works).

How many paths are created while rendering an animation?  Are most of these paths typically GCed around the same time?

Reply:
In the rive example app, I'm seeing about 100 paths per frame. It's not entirely clear to me about the timing of the GC.

I'll look at writing a benchmark for this.

Reply:
I wrote a benchmark.

I went with this strategy:

- `Erase()` now adds paths to a separate collection, guarded by a mutex
- `OnFrame()` calls `Drain()` before processing paths
- `Drain()` checks whether that collection needs to be drained. If so, it swaps it with an empty collection using the mutex, and then removes those paths from the set of paths we're tracking.
- OnFrame goes on to process any remaining paths.

The benchmark is now silly fast for this, whereas before it was relatively slow.